### DISCUSSION TITLE
BM3D Opencl

### DISCUSSION THREAD

**Discussion by ZenKiyoshi** [WolframRhodium/VapourSynth-BM3DCUDA#18] [General]
Is it possible to make BM3D opencl version, which can runs on both AMD and NVIDIA?

**Comment by WolframRhodium** [comment]
The opencl-c language lacks many crucial features for the current implementation. AMD support will be added when the ROCm software stack matures.

There is a good [opencl-accelerated BM3D](https://link.springer.com/article/10.1007/s11554-020-00945-4) in the literature, but I have no access to the code and I do not completely understand how to implement it.

**Comment by ZenKiyoshi** [comment]
[KNLMeansCL](https://github.com/Khanattila/KNLMeansCL) is an opencl implementation example, but it may be implemented in very different way.

It's may be helpful if you ask the author: https://forum.doom9.org/showthread.php?t=171379

As an AMD user, I sitll waiting for opencl implementation.

**Comment by WolframRhodium** [comment]
KNLMeansCL does not require these fancy features (in-lane permutations). Without them, it is measured to be as slow as 0.5x of the current implementation on my hardware. This performance can be inferior to the cpu implementation.

**Comment by ZenKiyoshi** [comment]
Thank for replying. So, ROCm is now the only help. Sorry, I am not familiar with software development, I don't know how it going on.

I don't use BM3D much currently, due to it's speed. When you are able to implement on AMD GPU, I will try it definitely.
